pkg/operator/controller/canary/daemonset.go

```diff
@@ -19,16 +23,21 @@ func (r *reconciler) ensureCanaryDaemonSet() (bool, *appsv1.DaemonSet, error) {
 		return false, nil, err
 	}
 
-	if haveDs {
-		return true, current, nil
-	}
-
-	err = r.createCanaryDaemonSet(desired)
-	if err != nil {
-		return false, nil, err
+	switch {
+	case !haveDs:
+		if err := r.createCanaryDaemonSet(desired); err != nil {
+			return false, nil, err
+		}
+		return r.currentCanaryDaemonSet()
+	case haveDs:
+		if updated, err := r.updateCanaryDaemonSet(current, desired); err != nil {
+			return true, current, err
+		} else if updated {
+			return r.currentCanaryDaemonSet()
+		}
 	}
 
-	return true, desired, nil
+	return true, current, nil
```

Comment:
Couldn't you `return r.currentCanaryDaemonSet()` here too? Then you wouldn't need them on lines 31 and 36.

Reply:
We use this flow to optimize out a few API calls.... we only need to fetch the current resource from the API if it were recently modified or created. If no change at all happens (ie no update necessary), then we can return `current` as is.